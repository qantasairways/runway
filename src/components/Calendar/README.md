```js
<Calendar
  isSelectingStartDate={true}
  departPlaceholder="Depart when?"
  returnPlaceholder="Return when?"
  departSelectedLabel="Depart"
  returnSelectedLabel="Return"
  returnDate="Tue 20 Mar 2019"
  departDate=""
  isOneWay={false}
/>
```
