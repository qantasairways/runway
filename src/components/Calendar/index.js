/* eslint-disable */
import React, { Component } from 'react';
import CalendarHeader from './components/CalendarHeader';

class Calendar extends Component {
  render = () => {
    return (
      <div>
        <CalendarHeader {...this.props} />
      </div>
    );
  };
}

export default Calendar;
