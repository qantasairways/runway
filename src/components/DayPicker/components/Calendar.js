import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { DayPicker } from 'react-day-picker';
import { css } from 'emotion';

import noop from '../../../utils/noop';
import { colours, fontFamily, fontSize } from '../../../theme/airways';

import Day from './Day';

function focusStyles() {
  return css({
    '&:focus': {
      zIndex: 2
    }
  });
}

function weekStyles(rowStyles) {
  return css({
    ...rowStyles
  });
}

function monthStyles() {
  return css({
    paddingTop: '2px',
    fontFamily: fontFamily.body,
    textAlign: 'center',
    background: colours.disabledGrey
  });
}

export function getDateToFocus({ start, disabledBefore }) {
  if (start) {
    return 'start';
  }

  if (disabledBefore) {
    return 'disabledBefore';
  }

  return 'today';
}

class Calendar extends Component {
  renderDay = (day, { hidden, ...modifiers }) => {
    const {
      setFocusElementRef,
      startLabel,
      endLabel,
      startAriaLabel,
      endAriaLabel,
      Icon
    } = this.props;

    return hidden ? null : (
      <Day
        setFocusElementRef={setFocusElementRef}
        day={day}
        modifiers={modifiers}
        startLabel={startLabel}
        endLabel={endLabel}
        startAriaLabel={startAriaLabel}
        endAriaLabel={endAriaLabel}
        Icon={Icon}
      />
    );
  };

  renderMonthCaption = ({ date }) => {
    const { monthLabels } = this.props;

    return (
      <div
        data-date={date}
        css={{
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'center',
          height: '58px'
        }}
      >
        <span
          css={{
            fontFamily: fontFamily.bold,
            fontSize: '1.375rem',
            marginRight: '5px'
          }}
        >
          {monthLabels[date.getMonth()]}
        </span>
        <span
          css={{
            fontSize: fontSize.body,
            fontFamily: fontFamily.body
          }}
        >
          {date.getFullYear()}
        </span>
      </div>
    );
  };

  render() {
    const {
      today,
      start,
      end,
      onDayClick,
      disabledBefore,
      hiddenBefore,
      disabledAfter,
      rowStyles,
      firstDayOfWeek,
      monthsToShow
    } = this.props;

    const focusElement = getDateToFocus({
      start,
      disabledBefore
    });

    const initialMonth = today;

    return (
      <DayPicker
        initialMonth={new Date(initialMonth.setHours(0, 0, 0, 0))}
        disabledDays={{ before: disabledBefore, after: disabledAfter }}
        selectedDays={{ from: start, to: end }}
        modifiers={{
          start,
          end,
          hidden: { before: hiddenBefore },
          focusElement: this.props[focusElement]
        }}
        renderDay={this.renderDay}
        captionElement={this.renderMonthCaption}
        onDayClick={onDayClick}
        numberOfMonths={monthsToShow}
        firstDayOfWeek={firstDayOfWeek}
        showWeekDays={false}
        canChangeMonth={false}
        modifiersStyles={{
          hidden: { display: 'none' }
        }}
        classNames={{
          body: 'runway-calendar__body',
          week: weekStyles(rowStyles),
          day: focusStyles(),
          month: monthStyles(),
          outside: 'outside',
          selected: 'selected',
          disabled: 'disabled',
          today: 'today'
        }}
      />
    );
  }
}

Calendar.propTypes = {
  today: PropTypes.instanceOf(Date),
  start: PropTypes.instanceOf(Date),
  end: PropTypes.instanceOf(Date),
  hiddenBefore: PropTypes.instanceOf(Date),
  disabledBefore: PropTypes.instanceOf(Date),
  disabledAfter: PropTypes.instanceOf(Date),
  onDayClick: PropTypes.func,
  setFocusElementRef: PropTypes.func,
  firstDayOfWeek: PropTypes.number,
  monthsToShow: PropTypes.number,
  startLabel: PropTypes.string,
  endLabel: PropTypes.string,
  startAriaLabel: PropTypes.string,
  endAriaLabel: PropTypes.string,
  monthLabels: PropTypes.arrayOf(PropTypes.string),
  Icon: PropTypes.func,
  rowStyles: PropTypes.shape()
};

Calendar.defaultProps = {
  today: new Date(),
  start: null,
  end: null,
  hiddenBefore: null,
  disabledBefore: null,
  disabledAfter: null,
  onDayClick: noop,
  setFocusElementRef: noop,
  firstDayOfWeek: 0,
  monthsToShow: 1,
  startLabel: '',
  endLabel: '',
  startAriaLabel: '',
  endAriaLabel: '',
  monthLabels: [],
  Icon: null,
  rowStyles: {}
};

export default Calendar;
