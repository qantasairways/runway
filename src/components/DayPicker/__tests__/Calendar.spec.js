import React from 'react';
import { shallow } from 'enzyme';
import Calendar, { getDateToFocus } from '../components/Calendar';

const CalendarIcon = () => <div>Calendar Icon</div>;

const today = new Date(2019, 3, 21, 0, 0, 0, 0);
const end = new Date(2019, 3, 24, 0, 0, 0, 0);
const disabledBefore = new Date(2019, 3, 10, 0, 0, 0, 0);
const hiddenBefore = new Date(2019, 3, 8, 0, 0, 0, 0);
const disabledAfter = new Date(2019, 4, 21, 0, 0, 0, 0);
global.Date = jest.fn(() => today);

const props = {
  start: today,
  end,
  hiddenBefore,
  disabledBefore,
  disabledAfter,
  onDayClick: e => e,
  setFocusElementRef: e => e,
  firstDayOfWeek: 4,
  monthsToShow: 6,
  startLabel: 'Depart',
  endLabel: 'Return',
  startAriaLabel: 'Selected for departure',
  endAriaLabel: 'Selected for return',
  monthLabels: ['Jan', 'Feb'],
  Icon: CalendarIcon,
  rowStyles: {}
};

describe('Calendar', () => {
  let component;

  it('renders correctly with defaults', () => {
    component = shallow(<Calendar today={new Date()} />);

    expect(component).toMatchSnapshot();
  });

  it('renders correctly with props provided', () => {
    component = shallow(<Calendar {...props} />);

    expect(component).toMatchSnapshot();
  });

  describe('renderDay()', () => {
    it('renders correctly', () => {
      const date = new Date(2019, 3, 10, 0, 0, 0, 0);
      const modifiers = { today: true };

      component = shallow(<Calendar {...props} />);

      const renderedDay = shallow(
        component.instance().renderDay(date, modifiers)
      );

      expect(renderedDay).toMatchSnapshot();
    });

    it('renders correctly when day should be hidden', () => {
      const date = new Date(2019, 3, 10, 0, 0, 0, 0);
      const modifiers = { hidden: true };

      component = shallow(<Calendar {...props} />);

      expect(component.instance().renderDay(date, modifiers)).toBe(null);
    });
  });

  describe('renderMonthCaption()', () => {
    it('renders correctly', () => {
      component = shallow(<Calendar {...props} />);

      const renderedMonthCaption = shallow(
        component.instance().renderMonthCaption({ date: new Date(2019, 12, 1) })
      );

      expect(renderedMonthCaption).toMatchSnapshot();
    });
  });

  describe('getDateToFocus()', () => {
    it('returns "start" if start date is selected', () => {
      expect(getDateToFocus({ start: disabledAfter, disabledBefore })).toBe(
        'start'
      );
    });

    it('returns the "disabledBefore" date if no start date is selected', () => {
      expect(getDateToFocus({ start: null, disabledBefore })).toBe(
        'disabledBefore'
      );
    });

    it('returns "today" if start date and disabledBefore date are not selected', () => {
      expect(getDateToFocus({ start: null, disabledBefore: null })).toBe(
        'today'
      );
    });
  });
});
