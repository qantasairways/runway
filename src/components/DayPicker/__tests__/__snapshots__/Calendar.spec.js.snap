// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar renderDay() renders correctly 1`] = `
.emotion-1 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  padding-top: 38px;
  box-sizing: border-box;
  height: 94px;
  color: #323232;
  background: #fff;
  box-shadow: 0 0 0 1px #eaeaea;
}

.emotion-0 {
  height: 1.125rem;
  padding: 0 4px;
  font-size: 1.125rem;
  line-height: 1.1;
  border-radius: 4px;
  background-color: #bff4f2;
}

<div
  aria-label=", "
  className="emotion-1"
>
  <div
    className="emotion-0"
  >
    21
  </div>
</div>
`;

exports[`Calendar renderMonthCaption() renders correctly 1`] = `
.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 58px;
}

.emotion-0 {
  font-family: "Ciutadella-Md",sans-serif;
  font-size: 1.375rem;
  margin-right: 5px;
}

.emotion-1 {
  font-size: 1.125rem;
  font-family: "Ciutadella",sans-serif;
}

<div
  className="emotion-2"
  data-date={2019-04-20T14:00:00.000Z}
>
  <span
    className="emotion-0"
  />
  <span
    className="emotion-1"
  >
    2019
  </span>
</div>
`;

exports[`Calendar renders correctly with defaults 1`] = `
<DayPicker
  canChangeMonth={false}
  captionElement={[Function]}
  classNames={
    Object {
      "body": "runway-calendar__body",
      "day": "css-76tx06",
      "disabled": "disabled",
      "month": "css-14fkfuc",
      "outside": "outside",
      "selected": "selected",
      "today": "today",
      "week": "css-0",
    }
  }
  disabledDays={
    Object {
      "after": null,
      "before": null,
    }
  }
  enableOutsideDaysClick={true}
  firstDayOfWeek={0}
  fixedWeeks={false}
  initialMonth={2019-04-20T14:00:00.000Z}
  labels={
    Object {
      "nextMonth": "Next Month",
      "previousMonth": "Previous Month",
    }
  }
  locale="en"
  localeUtils={
    Object {
      "default": Object {
        "formatDay": [Function],
        "formatMonthTitle": [Function],
        "formatWeekdayLong": [Function],
        "formatWeekdayShort": [Function],
        "getFirstDayOfWeek": [Function],
        "getMonths": [Function],
      },
      "formatDay": [Function],
      "formatMonthTitle": [Function],
      "formatWeekdayLong": [Function],
      "formatWeekdayShort": [Function],
      "getFirstDayOfWeek": [Function],
      "getMonths": [Function],
    }
  }
  modifiers={
    Object {
      "end": null,
      "focusElement": 2019-04-20T14:00:00.000Z,
      "hidden": Object {
        "before": null,
      },
      "start": null,
    }
  }
  modifiersStyles={
    Object {
      "hidden": Object {
        "display": "none",
      },
    }
  }
  navbarElement={
    <Navbar
      classNames={
        Object {
          "body": "DayPicker-Body",
          "caption": "DayPicker-Caption",
          "container": "DayPicker",
          "day": "DayPicker-Day",
          "disabled": "disabled",
          "footer": "DayPicker-Footer",
          "interactionDisabled": "DayPicker--interactionDisabled",
          "month": "DayPicker-Month",
          "months": "DayPicker-Months",
          "navBar": "DayPicker-NavBar",
          "navButtonInteractionDisabled": "DayPicker-NavButton--interactionDisabled",
          "navButtonNext": "DayPicker-NavButton DayPicker-NavButton--next",
          "navButtonPrev": "DayPicker-NavButton DayPicker-NavButton--prev",
          "outside": "outside",
          "selected": "selected",
          "today": "today",
          "todayButton": "DayPicker-TodayButton",
          "week": "DayPicker-Week",
          "weekNumber": "DayPicker-WeekNumber",
          "weekday": "DayPicker-Weekday",
          "weekdays": "DayPicker-Weekdays",
          "weekdaysRow": "DayPicker-WeekdaysRow",
          "wrapper": "DayPicker-wrapper",
        }
      }
      dir="ltr"
      labels={
        Object {
          "nextMonth": "Next Month",
          "previousMonth": "Previous Month",
        }
      }
      showNextButton={true}
      showPreviousButton={true}
    />
  }
  numberOfMonths={1}
  onDayClick={[Function]}
  pagedNavigation={false}
  renderDay={[Function]}
  renderWeek={[Function]}
  reverseMonths={false}
  selectedDays={
    Object {
      "from": null,
      "to": null,
    }
  }
  showOutsideDays={false}
  showWeekDays={false}
  showWeekNumbers={false}
  tabIndex={0}
  weekdayElement={<Weekday />}
/>
`;

exports[`Calendar renders correctly with props provided 1`] = `
<DayPicker
  canChangeMonth={false}
  captionElement={[Function]}
  classNames={
    Object {
      "body": "runway-calendar__body",
      "day": "css-76tx06",
      "disabled": "disabled",
      "month": "css-14fkfuc",
      "outside": "outside",
      "selected": "selected",
      "today": "today",
      "week": "css-0",
    }
  }
  disabledDays={
    Object {
      "after": 2019-05-20T14:00:00.000Z,
      "before": 2019-04-09T14:00:00.000Z,
    }
  }
  enableOutsideDaysClick={true}
  firstDayOfWeek={4}
  fixedWeeks={false}
  initialMonth={2019-04-20T14:00:00.000Z}
  labels={
    Object {
      "nextMonth": "Next Month",
      "previousMonth": "Previous Month",
    }
  }
  locale="en"
  localeUtils={
    Object {
      "default": Object {
        "formatDay": [Function],
        "formatMonthTitle": [Function],
        "formatWeekdayLong": [Function],
        "formatWeekdayShort": [Function],
        "getFirstDayOfWeek": [Function],
        "getMonths": [Function],
      },
      "formatDay": [Function],
      "formatMonthTitle": [Function],
      "formatWeekdayLong": [Function],
      "formatWeekdayShort": [Function],
      "getFirstDayOfWeek": [Function],
      "getMonths": [Function],
    }
  }
  modifiers={
    Object {
      "end": 2019-04-23T14:00:00.000Z,
      "focusElement": 2019-04-20T14:00:00.000Z,
      "hidden": Object {
        "before": 2019-04-07T14:00:00.000Z,
      },
      "start": 2019-04-20T14:00:00.000Z,
    }
  }
  modifiersStyles={
    Object {
      "hidden": Object {
        "display": "none",
      },
    }
  }
  navbarElement={
    <Navbar
      classNames={
        Object {
          "body": "DayPicker-Body",
          "caption": "DayPicker-Caption",
          "container": "DayPicker",
          "day": "DayPicker-Day",
          "disabled": "disabled",
          "footer": "DayPicker-Footer",
          "interactionDisabled": "DayPicker--interactionDisabled",
          "month": "DayPicker-Month",
          "months": "DayPicker-Months",
          "navBar": "DayPicker-NavBar",
          "navButtonInteractionDisabled": "DayPicker-NavButton--interactionDisabled",
          "navButtonNext": "DayPicker-NavButton DayPicker-NavButton--next",
          "navButtonPrev": "DayPicker-NavButton DayPicker-NavButton--prev",
          "outside": "outside",
          "selected": "selected",
          "today": "today",
          "todayButton": "DayPicker-TodayButton",
          "week": "DayPicker-Week",
          "weekNumber": "DayPicker-WeekNumber",
          "weekday": "DayPicker-Weekday",
          "weekdays": "DayPicker-Weekdays",
          "weekdaysRow": "DayPicker-WeekdaysRow",
          "wrapper": "DayPicker-wrapper",
        }
      }
      dir="ltr"
      labels={
        Object {
          "nextMonth": "Next Month",
          "previousMonth": "Previous Month",
        }
      }
      showNextButton={true}
      showPreviousButton={true}
    />
  }
  numberOfMonths={6}
  onDayClick={[Function]}
  pagedNavigation={false}
  renderDay={[Function]}
  renderWeek={[Function]}
  reverseMonths={false}
  selectedDays={
    Object {
      "from": 2019-04-20T14:00:00.000Z,
      "to": 2019-04-23T14:00:00.000Z,
    }
  }
  showOutsideDays={false}
  showWeekDays={false}
  showWeekNumbers={false}
  tabIndex={0}
  weekdayElement={<Weekday />}
/>
`;
